File: sdks/android/digm-sdk/digm-sdk/src/main/java/org/usexfg/digm/sdk/DigmSigner.kt
package org.usexfg.digm.sdk

import android.content.Context
import android.security.keystore.KeyGenParameterSpec
import android.security.keystore.KeyProperties
import java.security.KeyPairGenerator
import java.security.KeyStore
import java.util.Base64

data class DigmSignatureBundle(
    val sha256Base64: String,
    val signatureBase64: String,
    val publicKeyBase64: String,
    val nonce: String,
    val timestampMs: Long
)

class DigmSigner(private val context: Context) {
    private val alias = "digm_sign_key"

    private fun ensureKey() {
        val ks = KeyStore.getInstance("AndroidKeyStore").apply { load(null) }
        if (ks.containsAlias(alias)) return
        val kpg = KeyPairGenerator.getInstance(KeyProperties.KEY_ALGORITHM_EC, "AndroidKeyStore")
        val spec = KeyGenParameterSpec.Builder(
            alias,
            KeyProperties.PURPOSE_SIGN
        )
            .setDigests(KeyProperties.DIGEST_SHA256)
            .setUserAuthenticationRequired(false)
            .build()
        kpg.initialize(spec)
        kpg.generateKeyPair()
    }

    fun sign(sha256: ByteArray, nonce: String = java.util.UUID.randomUUID().toString()): DigmSignatureBundle {
        ensureKey()
        val ks = KeyStore.getInstance("AndroidKeyStore").apply { load(null) }
        val entry = ks.getEntry(alias, null) as KeyStore.PrivateKeyEntry
        val privateKey = entry.privateKey
        val publicKey = entry.certificate.publicKey.encoded

        val sig = java.security.Signature.getInstance("SHA256withECDSA")
        sig.initSign(privateKey)
        sig.update(sha256)
        val signature = sig.sign()

        val b64 = Base64.getEncoder()
        return DigmSignatureBundle(
            sha256Base64 = b64.encodeToString(sha256),
            signatureBase64 = b64.encodeToString(signature),
            publicKeyBase64 = b64.encodeToString(publicKey),
            nonce = nonce,
            timestampMs = System.currentTimeMillis()
        )
    }
}

File: sdks/ios/DIGMSDK/Sources/DIGMSDK/DigmSigner.swift
import Foundation
import CryptoKit

public struct DigmSignatureBundle: Codable {
    public let sha256Base64: String
    public let signatureBase64: String
    public let publicKeyBase64: String
    public let nonce: String
    public let timestampISO8601: String

    public init(sha256: Data, signature: Data, publicKey: Data, nonce: String) {
        self.sha256Base64 = sha256.base64EncodedString()
        self.signatureBase64 = signature.base64EncodedString()
        self.publicKeyBase64 = publicKey.base64EncodedString()
        self.nonce = nonce
        self.timestampISO8601 = ISO8601DateFormatter().string(from: Date())
    }
}

public final class DigmSigner {
    private var key: SecureEnclave.P256.Signing.PrivateKey?

    public init() {}

    private func getKey() throws -> SecureEnclave.P256.Signing.PrivateKey {
        if let k = key { return k }
        let k = try SecureEnclave.P256.Signing.PrivateKey()
        key = k
        return k
    }

    public func sign(sha256: Data, nonce: String = UUID().uuidString) throws -> DigmSignatureBundle {
        let signature = try getKey().signature(for: sha256)
        let pub = try getKey().publicKey.rawRepresentation
        return DigmSignatureBundle(
            sha256: sha256,
            signature: signature.derRepresentation,
            publicKey: pub,
            nonce: nonce
        )
    }
}
